client/state/terms/actions.js

```diff
@@ -45,7 +45,6 @@ export function addTerm( siteId, taxonomy, term ) {
 					dispatch( receiveTerm( siteId, taxonomy, data ) );
 					return data;
 				},
-				() => Promise.resolve() // Silently ignore failure so we can proceed to remove temporary
```

Comment:
After removing this, I can see the error notice, which is expected,

However, I can also see the category appears created, which is unexpected - the category actually wasn't created, and after a refresh, it will disappear.

Reply:
I fear that I might be a little out of my depth here. One way to solve this would be something like this - the logic being that the `meta` key contains the endpoint which is only created when the category is successfully created.

```js
const validTerms = terms.filter( term => term.meta );
```

From there, we could ensure that only validTerms are shown in the list under `blocks/taxonomy-manager`, but I'm not sure if that's too much of a workaround.

I'd appreciate your thoughts on handling this. :)

Reply:
I'd be happy to spend some more time and help you with this, but I'm going to be away for the next week. 

When I'm back, I'll be happy to take a look and understand the issue before offering a suggestion. 

In the meantime, if anyone from @Automattic/team-calypso has a spare cycle to take a look, that would be awesome.

Reply:
No problem, thanks again for the detailed look at this, and enjoy your time off next week!

Reply:
> However, I can also see the category appears created, which is unexpected - the category actually wasn't created, and after a refresh, it will disappear.

I took another look today - seems like we're optimistically creating the categories, but don't revert them if they don't get created successfully.

The way I see it, there are 2 possible solutions:

A: Removing the optimistic category creation and leaving only the realtime one:

```diff
diff --git a/client/state/terms/actions.js b/client/state/terms/actions.js
index 4d2ebcef3dc..24892ad4c19 100644
--- a/client/state/terms/actions.js
+++ b/client/state/terms/actions.js
@@ -26,15 +26,6 @@ import 'calypso/state/terms/init';
  */
 export function addTerm( siteId, taxonomy, term ) {
        return ( dispatch ) => {
-               const temporaryId = 'temporary' + uuid();
-
-               dispatch(
-                       receiveTerm( siteId, taxonomy, {
-                               ...term,
-                               ID: temporaryId,
-                       } )
-               );
-
                return wpcom
                        .site( siteId )
                        .taxonomy( taxonomy )
@@ -46,10 +37,6 @@ export function addTerm( siteId, taxonomy, term ) {
                                        return data;
                                },
                        )
-                       .then( ( data ) => {
-                               dispatch( removeTerm( siteId, taxonomy, temporaryId ) );
-                               return data;
-                       } );
        };
 }
```

B: Always removing the temporary one, rather than on success only:

```
diff --git a/client/state/terms/actions.js b/client/state/terms/actions.js
index 4d2ebcef3dc..7abb7a887a8 100644
--- a/client/state/terms/actions.js
+++ b/client/state/terms/actions.js
@@ -40,13 +40,11 @@ export function addTerm( siteId, taxonomy, term ) {
                        .taxonomy( taxonomy )
                        .term()
                        .add( term )
-                       .then(
-                               ( data ) => {
-                                       dispatch( receiveTerm( siteId, taxonomy, data ) );
-                                       return data;
-                               },
-                       )
                        .then( ( data ) => {
+                               dispatch( receiveTerm( siteId, taxonomy, data ) );
+                               return data;
+                       } )
+                       .finally( ( data ) => {
                                dispatch( removeTerm( siteId, taxonomy, temporaryId ) );
                                return data;
                        } );
```

I think I lean towards A, since it's cleaner and simpler - the optimistically created category is hidden behind the modal window, so I'd argue that it's just unnecessary.

Let me know if that helps!

Reply:
So sorry for the delay - been a super busy week, so haven't been able to get round to this yet. 

That's super helpful! Thanks for taking such a detailed look, and I'm surprised that I missed it. I've gone with Option A as you suggested and removed the temporarily created category - not exactly sure what it might have been achieving to start with? However, it doesn't seem to cause any regressions, and I can't find any insight on why it was included.

Let me know what you think!